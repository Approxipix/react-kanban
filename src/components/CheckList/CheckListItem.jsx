import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { updateCheckListItem, deleteCheckListItem } from '../../redux/boardReducer/actions';
import styled from 'styled-components'
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { CheckBox } from '../BaseComponent';

const Item = styled.div`
  position: relative;
  display: flex;
  margin-bottom: 10px;
  line-height: 36px;
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
`;

const Desc = styled.label`
  display: flex;
  align-items: center;
  width: 100%;
  padding-right: 40px;
  cursor: pointer;
`;

const Text = styled.div`
  font-size: 16px;
  cursor: pointer;
  text-decoration: ${props => props.status ? 'line-through' : 'none'}
  color: ${props => props.status ? '#ccc' : 'initial'}
  font-style: ${props => props.status ? 'italic' : 'normal'}
`;

const Check = styled.input`
  display: none;
`;


const Button = styled.button`
  position: absolute;
  right: 0;
  top: 50%;
  transform: translateY(-50%);
  padding: 7px;
  color: #afafaf;
  line-height: 0;
  border: 0;
  background-color: transparent;
  outline: none;
  cursor: pointer;
  transition: all 0.2s ease-in;
  
  &:hover {
    color: #ccc;
  }
`;


class CheckListItem extends Component {

  handleDelete = () => {
    this.props.actions.deleteCheckListItem({
      cardIndex: this.props.cardIndex,
      checkBoxItemIndex: this.props.checkBoxItemIndex,
      listIndex: this.props.listIndex,
      checkBoxIndex: this.props.checkBoxIndex,
    })
  };

  handleStatus = () => {
    this.props.actions.updateCheckListItem({
      cardIndex: this.props.cardIndex,
      checkBoxItemIndex: this.props.checkBoxItemIndex,
      listIndex: this.props.listIndex,
      checkBoxIndex: this.props.checkBoxIndex,
      status: !this.props.status,
    })
  };

  render() {
    const { status, description } = this.props;
    if (this.props.hide && status) return null;
    return (
      <Item>
        <Desc onClick={(e) => e.stopPropagation() }>
          <CheckBox status={status}>
            <span>
              <svg width="12px" height="10px" viewBox="0 0 12 10">
                <polyline points="1.5 6 4.5 9 10.5 1" />
              </svg>
            </span>
          </CheckBox>
          <Check
            type="checkbox"
            onChange={() => this.handleStatus()}
            checked={status}
          />
          <Text status={status}>{description}</Text>
        </Desc>
        <Button type="button"
                onClick={() => this.handleDelete()}
        >
          delete
        </Button>
      </Item>
    );
  }
}

function mapStateToProps(state, ownProps) {
  const { rootReducer } = state;
  const board = rootReducer.boards[rootReducer.currentBoardIndex];
  const list = board.lists[ownProps.listIndex];
  const card = list.cards[ownProps.cardIndex];
  const items = card.checkLists[ownProps.checkBoxIndex].items[ownProps.checkBoxItemIndex];
  const status = card.checkLists[ownProps.checkBoxIndex].items[ownProps.checkBoxItemIndex].status;
  const description = card.checkLists[ownProps.checkBoxIndex].items[ownProps.checkBoxItemIndex].description;

  return {
    items: items,
    status: status,
    description: description
  }

}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      updateCheckListItem: updateCheckListItem,
      deleteCheckListItem: deleteCheckListItem,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(CheckListItem);
