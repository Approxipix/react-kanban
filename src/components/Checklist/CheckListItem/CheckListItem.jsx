import React, { Component } from 'react';
import PropTypes from "prop-types";
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { updateCheckListItem, deleteCheckListItem } from '../../../redux/checkListReducer/actions/actions';
import { CheckBox } from '../../BaseComponent';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components';

const HideCheckBox = styled.input`
  display: none;
`;

const Description = styled.div`
  font-size: .875rem;
  color: ${props => props.status ? '#6b778c' : 'initial'}
  font-style: ${props => props.status ? 'italic' : 'normal'}
  text-decoration: ${props => props.status ? 'line-through' : 'none'}
`;

const Button = styled.button`
  position: absolute;
  left: 100%;
  top: 50%;
  font-size: 1rem;
  color: #e44979;
  cursor: pointer;
  transition: all 0.2s ease-in;
  transform: translate(0, -50%);
`;
Button.displayName = 'Button';

const Wrapper = styled.label`
  position: relative;
  display: flex;
  padding: .5rem;
  border-radius: .2rem;
  transition: background-color .2s ease-in;
  cursor: pointer;
  overflow: hidden;
  &:hover {
    background-color: rgba(9, 30, 66, .08);
    ${Button} {
      transform: translate(-105%, -50%);
    }
  }
`;

class CheckListItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      status: props.status || false,
    };
  }

  componentDidUpdate(prevProps) {
    if (prevProps.status !== this.props.status) {
      this.setState({ status: this.props.status });
    }
  }

  handleDelete = () => {
    this.props.actions.deleteCheckListItem({
      checkListId: this.props.checkListId,
      checkListItemIndex: this.props.index,
    })
  };

  handleStatus = () => {
    this.setState({
      status: !this.state.status
    }, () => {
      this.props.actions.updateCheckListItem({
        checkListId: this.props.checkListId,
        checkListItemIndex: this.props.index,
        status: this.state.status,
      })
    })
  };

  render() {
    const { description, hideCompletedItems } = this.props;
    const { status } = this.state;
    if (hideCompletedItems && status) return null;
    return (
      <Wrapper onClick={(e) => e.stopPropagation()}>
        <CheckBox status={status}>
          <span>
            <svg width="10px" height="8px" viewBox="0 0 12 10">
              <polyline points="1.5 6 4.5 9 10.5 1" />
            </svg>
          </span>
        </CheckBox>
        <HideCheckBox
          type="checkbox"
          onChange={() => this.handleStatus()}
          checked={status}
        />
        <Description status={status}>
          {description}
        </Description>
        <Button type="button" onClick={() => this.handleDelete()}>
          <FontAwesomeIcon icon="times"/>
        </Button>
      </Wrapper>
    );
  }
}

CheckListItem.defaultProps = {
  status: false,
  description: '',
  hideCompletedItems: false,
};

CheckListItem.propTypes = {
  index: PropTypes.number.isRequired,
  checkListId: PropTypes.string.isRequired,
  status: PropTypes.bool.isRequired,
  description: PropTypes.string.isRequired,
  hideCompletedItems: PropTypes.bool.isRequired,
};

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      updateCheckListItem: updateCheckListItem,
      deleteCheckListItem: deleteCheckListItem,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CheckListItem);
