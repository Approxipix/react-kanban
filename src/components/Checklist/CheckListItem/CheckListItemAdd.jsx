import React, { Component } from "react";
import { connect } from "react-redux";
import styled from 'styled-components'
import { bindActionCreators } from 'redux';
import { addCheckListItem } from '../../../redux/checkListReducer/actions/actions';
import { Actions, Input, SubmitButton, CancelButton } from '../../BaseComponent';
import ClickOutside from "../../ClickOutside/ClickOutside";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Wrapper = styled.div`
  padding-left: 2rem;
`;

const AddButton = styled.button`
  padding: .5rem 1rem;
  font-size: .875rem;
  color: #172b4d;
  background-color: rgba(9, 30, 66, .08);
  border-radius: 3px;
  transition: background-color .2s ease-in;
  &:hover {
    background-color: rgba(9,30,66,.13);
  }
`;
AddButton.displayName = 'AddButton';

class CheckListItemAdd extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: false,
      title: '',
    };
  }

  toggleOpened = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      title: '',
    })
  };

  handleChange = (event) => {
    this.setState({
      title: event.target.value,
    });
  };

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.toggleOpened();
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    if (!title)  return;
    this.props.actions.addCheckListItem({
      checkListId: this.props.checkListId,
      checkListTitle: title,
    });
    this.toggleOpened()
  };

  render = () => {
    const { isOpened, title } = this.state;
    return isOpened ? (
      <ClickOutside toggleOpened={this.toggleOpened}>
        <Wrapper>
          <form>
            <Input
              type="text"
              size=".875rem"
              padding=".5rem"
              placeholder="Enter a card title..."
              value={title}
              onKeyDown={this.handleKeyDown}
              onChange={(e) => this.handleChange(e)}
              spellCheck={false}
              autoFocus
            />
            <Actions>
              <SubmitButton onClick={(e) => this.handleSubmit(e)} disabled={!title}>
                Save
              </SubmitButton>
              <CancelButton onClick={() => this.toggleOpened()}>
                <FontAwesomeIcon icon="times" />
              </CancelButton>
            </Actions>
          </form>
        </Wrapper>
      </ClickOutside>
    ) : (
      <Wrapper>
        <AddButton onClick={() => this.toggleOpened()}>
          Add a item
        </AddButton>
      </Wrapper>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addCheckListItem: addCheckListItem,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CheckListItemAdd);
