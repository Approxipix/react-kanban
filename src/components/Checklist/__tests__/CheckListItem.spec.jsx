import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CheckListItem from '../CheckListItem/CheckListItem';

const mockStore = configureStore([]);

describe('<CheckListItem>', () => {
  let store;
  let props;
  let component;
  let mockUpdateCheckListItem = jest.fn();
  let mockDeleteCheckListItem = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();

    props = {
      index: 0,
      checkListId: 'CheckListID',
      description: 'Description',
      hideCompletedItems: false,
    };
    const state = {
      status: false,
    };
    store = mockStore({});
    store.dispatch = mockDeleteCheckListItem;

    const wrapper = shallow(<CheckListItem store={store} {...props} />);
    component = wrapper.find('CheckListItem').dive();
    component.setState({...state});
  });

  it('should renders without crashing given the required props', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should return if hideCompletedItems and status equals true', () => {
    const props = {
      hideCompletedItems: true,
    };
    const state = {
      status: true,
    };

    const wrapper = shallow(<CheckListItem store={store} {...props} />);
    component = wrapper.find('CheckListItem').dive();
    component.setState({...state});

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should call handleStatus on checkbox change with the correct params', () => {
    const spy = jest.spyOn(component.instance(), "handleStatus");

    component.find({ type: 'checkbox' }).simulate("change");

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should call setState on status', () => {
    const expectedState = {
      status: true,
    };

    component.instance().handleStatus();

    expect(component.state()).toEqual(expectedState)
  });

  it('should dispatch an action on delete button click ', () => {
    const expectedAction = {
      type: 'DELETE_CHECKLIST_ITEM',
      payload: {
        checkListId: 'CheckListID',
        checkListItemIndex: 0,
      }
    };

    component.find('Button').simulate('click');

    expect(mockDeleteCheckListItem).toHaveBeenCalledTimes(1);
    expect(mockDeleteCheckListItem).toHaveBeenCalledWith(expectedAction);
  });

  it('should dispatch an action on handleStatus ', () => {
    const expectedAction = {
      type: 'UPDATE_CHECKLIST_ITEM',
      payload: {
        checkListId: 'CheckListID',
        checkListItemIndex: 0,
        status: true,
      }
    };
    store.dispatch = mockUpdateCheckListItem;
    const wrapper = shallow(<CheckListItem store={store} {...props} />);
    component = wrapper.find('CheckListItem').dive();

    component.instance().handleStatus();

    expect(mockUpdateCheckListItem).toHaveBeenCalledTimes(1);
    expect(mockUpdateCheckListItem).toHaveBeenCalledWith(expectedAction);
  });
});
