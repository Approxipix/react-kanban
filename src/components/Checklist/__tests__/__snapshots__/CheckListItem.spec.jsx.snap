// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckListItem> should renders without crashing given the required props 1`] = `
<styled.label
  onClick={[Function]}
>
  <styled.div
    status={false}
  >
    <span>
      <svg
        height="8px"
        viewBox="0 0 12 10"
        width="10px"
      >
        <polyline
          points="1.5 6 4.5 9 10.5 1"
        />
      </svg>
    </span>
  </styled.div>
  <styled.input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <styled.div
    status={false}
  >
    Description
  </styled.div>
  <Button
    onClick={[Function]}
    type="button"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="times"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </Button>
</styled.label>
`;

exports[`<CheckListItem> should return if hideCompletedItems and status equals true 1`] = `""`;
